File: HighFive/HighFive/System/Text/UnicodeEncoding.cs
namespace System.Text
{
    public class UnicodeEncoding : Encoding
    {
        private bool bigEndian;
        private bool byteOrderMark;
        private bool throwOnInvalid;

        public UnicodeEncoding() : this(false, true)
        {
        }

        public UnicodeEncoding(bool bigEndian, bool byteOrderMark) : this(bigEndian, byteOrderMark, false)
        {
        }

        public UnicodeEncoding(bool bigEndian, bool byteOrderMark, bool throwOnInvalidBytes)
        {
            this.bigEndian = bigEndian;
            this.byteOrderMark = byteOrderMark;
            this.throwOnInvalid = throwOnInvalidBytes;
            this.fallbackCharacter = '\uFFFD';
        }

        public override int CodePage => this.bigEndian ? 1201 : 1200;
        public override string EncodingName => this.bigEndian ? "Unicode (Big-Endian)" : "Unicode";

        protected override byte[] Encode(string s, byte[] outputBytes, int outputIndex, out int writtenBytes)
        {
            var hasBuffer = outputBytes != null;
            var recorded = 0;
            char surrogate_1st = '\u0000';
            var fallbackCharacterCode = this.fallbackCharacter;

            Action<byte> write = ch =>
            {
                if (hasBuffer)
                {
                    if (outputIndex >= outputBytes.Length)
                    {
                        throw new System.ArgumentException("bytes");
                    }

                    outputBytes[outputIndex++] = ch;
                }
                else
                {
                    outputBytes.Push(ch);
                }
                recorded++;
            };

            Action<byte, byte> writePair = (a, b) =>
            {
                write(a);
                write(b);
            };

            Func<char, char> swap = ch => (char)(((ch & 0xFF) << 8) | ((ch >> 8) & 0xFF));

            Action fallback = () =>
            {
                if (this.throwOnInvalid)
                {
                    throw new System.Exception("Invalid character in UTF16 text");
                }

                writePair((byte)fallbackCharacterCode, (byte)(fallbackCharacterCode >> 8));
            };

            if (!hasBuffer)
            {
                outputBytes = new byte[0];
            }

            if (this.bigEndian)
            {
                fallbackCharacterCode = swap(fallbackCharacterCode);
            }

            for (var i = 0; i < s.Length; i++)
            {
                var ch = s[i];

                if (surrogate_1st != 0)
                {
                    if (ch >= 0xDC00 && ch <= 0xDFFF)
                    {
                        if (this.bigEndian)
                        {
                            surrogate_1st = swap(surrogate_1st);
                            ch = swap(ch);
                        }
                        writePair((byte)surrogate_1st, (byte)(surrogate_1st >> 8));
                        writePair((byte)ch, (byte)(ch >> 8));
                        surrogate_1st = '\u0000';
                        continue;
                    }
                    fallback();
                    surrogate_1st = '\u0000';
                }

                if (0xD800 <= ch && ch <= 0xDBFF)
                {
                    surrogate_1st = ch;
                    continue;
                }
                else if (0xDC00 <= ch && ch <= 0xDFFF)
                {
                    fallback();
                    surrogate_1st = '\u0000';
                    continue;
                }

                if (ch < 0x10000)
                {
                    if (this.bigEndian)
                    {
                        ch = swap(ch);
                    }
                    writePair((byte)ch, (byte)(ch >> 8));
                }
                else if (ch <= 0x10FFFF)
                {
                    ch = H5.Script.Write<char>("ch - 0x10000"); //?????

                    char lowBits = (char)((ch & 0x3FF) | 0xDC00);
                    char highBits = (char)(((ch >> 10) & 0x3FF) | 0xD800);

                    if (this.bigEndian)
                    {
                        highBits = swap(highBits);
                        lowBits = swap(lowBits);
                    }
                    writePair((byte)highBits, (byte)(highBits >> 8));
                    writePair((byte)lowBits, (byte)(lowBits >> 8));
                }
                else
                {
                    fallback();
                }
            }

            if (surrogate_1st != 0)
            {
                fallback();
            }

            writtenBytes = recorded;

            if (hasBuffer)
            {
                return null;
            }

            return outputBytes;
        }

        protected override string Decode(byte[] bytes, int index, int count, char[] chars, int charIndex)
        {
            var position = index;
            var result = "";
            var endpoint = position + count;
            this._hasError = false;

            Action fallback = () =>
            {
                if (this.throwOnInvalid)
                {
                    throw new System.Exception("Invalid character in UTF16 text");
                }

                result += this.fallbackCharacter;
            };

            Func<char, char> swap = ch => (char)(((byte)ch << 8) | (byte)(ch >> 8));

            Func<char?> readPair = () =>
            {
                if ((position + 2) > endpoint)
                {
                    position = position + 2;
                    return null;
                }

                var a = bytes[position++];
                var b = bytes[position++];

                var point = (char)((a << 8) | b);
                if (!this.bigEndian)
                {
                    point = swap(point);
                }

                return point;
            };

            while (position < endpoint)
            {
                var firstWord = readPair();

                if (!firstWord.HasValue)
                {
                    fallback();
                    this._hasError = true;
                }
                else if ((firstWord < 0xD800) || (firstWord > 0xDFFF))
                {
                    result += Encoding.FromCharCode(firstWord.Value);
                }
                else if ((firstWord >= 0xD800) && (firstWord <= 0xDBFF))
                {
                    var end = position >= endpoint;
                    var secondWord = readPair();
                    if (end)
                    {
                        fallback();
                        this._hasError = true;
                    }
                    else if (!secondWord.HasValue)
                    {
                        fallback();
                        fallback();
                    }
                    else if ((secondWord >= 0xDC00) && (secondWord <= 0xDFFF))
                    {
                        var highBits = firstWord & 0x3FF;
                        var lowBits = secondWord & 0x3FF;

                        var charCode = ((highBits << 10) | lowBits) + 0x10000;

                        result += Encoding.FromCharCode(charCode.Value);
                    }
                    else
                    {
                        fallback();
                        position = position - 2;
                    }
                }
                else
                {
                    fallback();
                }
            }

            return result;
        }

        public override int GetMaxByteCount(int charCount)
        {
            if (charCount < 0)
            {
                throw new System.ArgumentOutOfRangeException("charCount");
            }

            var byteCount = (long)charCount + 1;
            byteCount <<= 1;

            if (byteCount > 0x7fffffff)
            {
                throw new System.ArgumentOutOfRangeException("charCount");
            }

            return (int)byteCount;
        }

        public override int GetMaxCharCount(int byteCount)
        {
            if (byteCount < 0)
            {
                throw new System.ArgumentOutOfRangeException("byteCount");
            }

            var charCount = (long)(byteCount >> 1) + (byteCount & 1) + 1;

            if (charCount > 0x7fffffff)
            {
                throw new System.ArgumentOutOfRangeException("byteCount");
            }

            return (int)charCount;
        }
    }
}

File: H5/H5/Resources/.generated/System/Text/UnicodeEncoding.js
    H5.define("System.Text.UnicodeEncoding", {
        inherits: [System.Text.Encoding],
        fields: {
            bigEndian: false,
            byteOrderMark: false,
            throwOnInvalid: false
        },
        props: {
            CodePage: {
                get: function () {
                    return this.bigEndian ? 1201 : 1200;
                }
            },
            EncodingName: {
                get: function () {
                    return this.bigEndian ? "Unicode (Big-Endian)" : "Unicode";
                }
            }
        },
        ctors: {
            ctor: function () {
                System.Text.UnicodeEncoding.$ctor1.call(this, false, true);
            },
            $ctor1: function (bigEndian, byteOrderMark) {
                System.Text.UnicodeEncoding.$ctor2.call(this, bigEndian, byteOrderMark, false);
            },
            $ctor2: function (bigEndian, byteOrderMark, throwOnInvalidBytes) {
                this.$initialize();
                System.Text.Encoding.ctor.call(this);
                this.bigEndian = bigEndian;
                this.byteOrderMark = byteOrderMark;
                this.throwOnInvalid = throwOnInvalidBytes;
                this.fallbackCharacter = 65533;
            }
        },
        methods: {
            Encode$3: function (s, outputBytes, outputIndex, writtenBytes) {
                var hasBuffer = outputBytes != null;
                var recorded = 0;
                var surrogate_1st = 0;
                var fallbackCharacterCode = this.fallbackCharacter;

                var write = function (ch) {
                    if (hasBuffer) {
                        if (outputIndex >= outputBytes.length) {
                            throw new System.ArgumentException.$ctor1("bytes");
                        }

                        outputBytes[System.Array.index(H5.identity(outputIndex, ((outputIndex = (outputIndex + 1) | 0))), outputBytes)] = ch;
                    } else {
                        outputBytes.push(ch);
                    }
                    recorded = (recorded + 1) | 0;
                };

                var writePair = function (a, b) {
                    write(a);
                    write(b);
                };

                var swap = $asm.$.System.Text.UnicodeEncoding.f1;

                var fallback = H5.fn.bind(this, function () {
                    if (this.throwOnInvalid) {
                        throw new System.Exception("Invalid character in UTF16 text");
                    }

                    writePair((fallbackCharacterCode & 255), ((fallbackCharacterCode >> 8) & 255));
                });

                if (!hasBuffer) {
                    outputBytes = System.Array.init(0, 0, System.Byte);
                }

                if (this.bigEndian) {
                    fallbackCharacterCode = swap(fallbackCharacterCode);
                }

                for (var i = 0; i < s.length; i = (i + 1) | 0) {
                    var ch = s.charCodeAt(i);

                    if (surrogate_1st !== 0) {
                        if (ch >= 56320 && ch <= 57343) {
                            if (this.bigEndian) {
                                surrogate_1st = swap(surrogate_1st);
                                ch = swap(ch);
                            }
                            writePair((surrogate_1st & 255), ((surrogate_1st >> 8) & 255));
                            writePair((ch & 255), ((ch >> 8) & 255));
                            surrogate_1st = 0;
                            continue;
                        }
                        fallback();
                        surrogate_1st = 0;
                    }

                    if (55296 <= ch && ch <= 56319) {
                        surrogate_1st = ch;
                        continue;
                    } else if (56320 <= ch && ch <= 57343) {
                        fallback();
                        surrogate_1st = 0;
                        continue;
                    }

                    if (ch < 65536) {
                        if (this.bigEndian) {
                            ch = swap(ch);
                        }
                        writePair((ch & 255), ((ch >> 8) & 255));
                    } else if (ch <= 1114111) {
                        ch = ch - 0x10000;

                        var lowBits = ((ch & 1023) | 56320) & 65535;
                        var highBits = (((ch >> 10) & 1023) | 55296) & 65535;

                        if (this.bigEndian) {
                            highBits = swap(highBits);
                            lowBits = swap(lowBits);
                        }
                        writePair((highBits & 255), ((highBits >> 8) & 255));
                        writePair((lowBits & 255), ((lowBits >> 8) & 255));
                    } else {
                        fallback();
                    }
                }

                if (surrogate_1st !== 0) {
                    fallback();
                }

                writtenBytes.v = recorded;

                if (hasBuffer) {
                    return null;
                }

                return outputBytes;
            },
            Decode$2: function (bytes, index, count, chars, charIndex) {
                var position = index;
                var result = "";
                var endpoint = (position + count) | 0;
                this._hasError = false;

                var fallback = H5.fn.bind(this, function () {
                    if (this.throwOnInvalid) {
                        throw new System.Exception("Invalid character in UTF16 text");
                    }

                    result = (result || "") + String.fromCharCode(this.fallbackCharacter);
                });

                var swap = $asm.$.System.Text.UnicodeEncoding.f2;

                var readPair = H5.fn.bind(this, function () {
                    if ((((position + 2) | 0)) > endpoint) {
                        position = (position + 2) | 0;
                        return null;
                    }

                    var a = bytes[System.Array.index(H5.identity(position, ((position = (position + 1) | 0))), bytes)];
                    var b = bytes[System.Array.index(H5.identity(position, ((position = (position + 1) | 0))), bytes)];

                    var point = ((a << 8) | b) & 65535;
                    if (!this.bigEndian) {
                        point = swap(point);
                    }

                    return point;
                });

                while (position < endpoint) {
                    var firstWord = readPair();

                    if (!System.Nullable.hasValue(firstWord)) {
                        fallback();
                        this._hasError = true;
                    } else if ((System.Nullable.lt(firstWord, 55296)) || (System.Nullable.gt(firstWord, 57343))) {
                        result = (result || "") + ((System.String.fromCharCode(System.Nullable.getValue(firstWord))) || "");
                    } else if ((System.Nullable.gte(firstWord, 55296)) && (System.Nullable.lte(firstWord, 56319))) {
                        var end = position >= endpoint;
                        var secondWord = readPair();
                        if (end) {
                            fallback();
                            this._hasError = true;
                        } else if (!System.Nullable.hasValue(secondWord)) {
                            fallback();
                            fallback();
                        } else if ((System.Nullable.gte(secondWord, 56320)) && (System.Nullable.lte(secondWord, 57343))) {
                            var highBits = System.Nullable.band(firstWord, 1023);
                            var lowBits = System.Nullable.band(secondWord, 1023);

                            var charCode = H5.Int.clip32(System.Nullable.add((System.Nullable.bor((System.Nullable.sl(highBits, 10)), lowBits)), 65536));

                            result = (result || "") + ((System.String.fromCharCode(System.Nullable.getValue(charCode))) || "");
                        } else {
                            fallback();
                            position = (position - 2) | 0;
                        }
                    } else {
                        fallback();
                    }
                }

                return result;
            },
            GetMaxByteCount: function (charCount) {
                if (charCount < 0) {
                    throw new System.ArgumentOutOfRangeException.$ctor1("charCount");
                }

                var byteCount = System.Int64(charCount).add(System.Int64(1));
                byteCount = byteCount.shl(1);

                if (byteCount.gt(System.Int64(2147483647))) {
                    throw new System.ArgumentOutOfRangeException.$ctor1("charCount");
                }

                return System.Int64.clip32(byteCount);
            },
            GetMaxCharCount: function (byteCount) {
                if (byteCount < 0) {
                    throw new System.ArgumentOutOfRangeException.$ctor1("byteCount");
                }

                var charCount = System.Int64((byteCount >> 1)).add(System.Int64((byteCount & 1))).add(System.Int64(1));

                if (charCount.gt(System.Int64(2147483647))) {
                    throw new System.ArgumentOutOfRangeException.$ctor1("byteCount");
                }

                return System.Int64.clip32(charCount);
            }
        }
    });

    H5.ns("System.Text.UnicodeEncoding", $asm.$);

    H5.apply($asm.$.System.Text.UnicodeEncoding, {
        f1: function (ch) {
            return ((((ch & 255) << 8) | ((ch >> 8) & 255)) & 65535);
        },
        f2: function (ch) {
            return ((((ch & 255) << 8) | (((ch >> 8)) & 255)) & 65535);
        }
    });
